src/track/track.cpp

```diff
@@ -160,50 +166,69 @@ void Track::importMetadata(
                     m_record.ptrMetadataSynchronized(),
                     true);
         }
-        bool modifiedReplayGain = false;
+
+        const auto oldReplayGain =
+                m_record.getMetadata().getTrackInfo().getReplayGain();
         if (m_record.getMetadata() != importedMetadata) {
-            modifiedReplayGain =
-                    (m_record.getMetadata().getTrackInfo().getReplayGain() != newReplayGain);
             m_record.setMetadata(std::move(importedMetadata));
             // Don't use importedMetadata after move assignment!!
             modified = true;
         }
-        if (modified) {
-            // explicitly unlock before emitting signals
-            markDirtyAndUnlock(&lock);
-            if (modifiedReplayGain) {
-                emit replayGainUpdated(newReplayGain);
-            }
+        const auto newReplayGain =
+                m_record.getMetadata().getTrackInfo().getReplayGain();
+
+        // Need to set BPM after sample rate since beat grid creation depends on
+        // knowing the sample rate. Bug #1020438.
+        const bool beatsAndBpmModified =
+                // Only use the imported BPM if the current beat grid is either
+                // missing or not valid! The BPM value in the metadata might be
+                // imprecise (normalized or rounded), e.g. ID3v2 only supports
+                // integer values.
+                !(m_pBeats && mixxx::Bpm::isValidValue(m_pBeats->getBpm())) &&
+                trySetBpmWhileLocked(importedBpm.getValue());
```

Comment:
Yes, this is valid c, but it I not obvious that the set in the last part is conditionally executed, when skimming over the code.
Please consider to use here a normal if condition.
The same applies to the code below.

Reply:
Ok. But then the flags become mutable and we need to use old, imperative style.

Reply:
Decide yourself. I would actually preferethat because a conditional block in { } it makes the control flow obvious.

Reply:
Already done.